Handle case sensitive file systems

doowb opened the pull request:
This is the code that @jonschlinkert and I worked through to provide a fallback for platforms like `linux` that are case sensitive.
- First try `fs.accessSync` or `fs.statSync`.
  - If exact match, these methods will succeed.
  - If case insensitive file system and file exists, these methods will succeed.
  - If case sensitive file system and file exists but different case, these methods will fail.
  - Returns given `filepath` on success (change from previous version which returned true).
- On case sensitive file systems (`linux`) fall back to checking entire directory for file.
  - Try to read `filepath` as a directory
  - Try to read `dirname(filepath)` as a directory
  - If `filepath` is a directory and read successful, return `filepath`
  - Iterate over files in directory and compare resolved filepaths to given filepath.
    - strict compare (same case)
    - lower cased filepath with directory filepath
    - given filepath with lowercased directory filepath
    - lower cased filepath with lowercased directory filepath
  - If any comparison matches, return matched filepath that was read in from the directory. This give consuming libraries a way to use the actual filepath with `fs` methods without having to redo all this searching and checks again.

doowb committed:
handle case sensitive file systems

- if the initial check does't succeed and the platform is linux, check against all the files in the same directory.
- update travis-ci config to run tests against osx and linux

doowb committed:
return early if an invalid filepath argument is passed in.

doowb committed:
paths are resolved on linux when returned

jonschlinkert commented:
@phated, this is what @doowb and I came up with to resolve the linux issue. We tried and discussed a number of different approaches, this one is the best we could come up with. 

If this seems okay to you, we can publish this change and update findup-sync and liftoff and let you know when it's finished. We're open to other approaches if anyone has any ideas.

phated commented on index.js:
```diff
@@ -8,11 +8,100 @@
+ *
+ * @name exists
+ * @param  {String} `filepath` filepath to check for.
+ * @return {String|Boolean} Returns the found filepath if it exists, otherwise returns `false`.
```
why return false instead of null/undefined?

jonschlinkert replied:
Isn't `false` what `fs.existSync` would return? When a path is found, we return the path since `true` wouldn't be a reliable result on case-insensitive file systems, since the user wouldn't know "which path" existed.

phated replied:
Yeah, but if you are changing the truthy behavior, it's no longer a replacement at all.  More reason to make this a library on top

jonschlinkert replied:
@doowb suggested we make `fs-resolve-file` or something like that to do this. that seems like a good idea. that way we can preserve the boolean behavior of this lib in all cases, instead of making it optional. sound ok to you?

phated replied:
I'm good with a separate package. I think the term "resolve" is usually reserved for require-like resolution in node land.  Maybe `lookup-file` or `find-file`

phated commented on index.js:
```diff
@@ -8,11 +8,100 @@
+    return filepath;
   } catch (err) {}
+
+  if (process.platform === 'linux') {
```
OSX can be installed with a case-sensitive filesystem and I think MacFS is going to be case sensitive.  Maybe this should always be called?  Or maybe it should be `!== 'win32'`.  Thoughts?

jonschlinkert replied:
okay, let do more research then. It's my understanding that only linux has this specific issue with `fs.statSync`. If anyone knows where this is documented that would be helpful

jonschlinkert replied:
I see what you mean about other file systems. Per my additional comments, I think a good solution is to just make the check optional, then we can keep this a drop-in replacement and allow the user to decide if the new logic will be used. 

Regardless of whether or not the logic is done here, I think this should be optional for the user. I personally would avoid using it if it was turned on by default, because it would be slow and I don't use a file system that needs the check.

phated commented on index.js:
```diff
@@ -8,11 +8,100 @@
+
+  while (++idx < len) {
+    var fp = path.resolve(res.path, res.files[idx]);
+    if (filepath === fp || lower === fp) {
```
is it more efficient to do these equality checks before doing `fp.toLowerCase` and then checking again?

jonschlinkert replied:
we could do `.toLowerCase()` inside the loop if the other check fails. we could just set the variable to undefined outside the loop and check if it's set so we only do `.toLowerCase()` once.

phated commented:
If these changes are going to land here (as opposed to being a separate library), the description, etc of this library will need to change since it is no longer a drop-in replacement.  Would also require a major bump.  What is the reason for incorporating into this library?

jonschlinkert commented:
> If these changes are going to land here (as opposed to being a separate library),

hmm, I must have misunderstood what the issue was.  It was my understanding that the logic used for checking if a file exists is unreliable on linux. 

I re-read the related issues to be sure, but if that's not the case, can you explain what it is that we're trying to fix? 

>  the description, etc of this library will need to change since it is no longer a drop-in replacement.

We can make the check optional and turn it off by default so that it would still be a drop-in replacement unless opted-in.

phated commented:
node defers all `fs` calls to native methods (libuv is just a shim that normalizes those calls to the native libraries, e.g. libc on linux & osx).  The issue isn't that fs-exists-sync (or fs.existsSync) is not operating correctly, but that we want to consistently support case-insensitivity in Liftoff (which means it needs to be supported in findup-sync).  I'm not sure "fixing" this in a shim is required and could be solved in a library on top of it or in findup-sync.

jonschlinkert commented:
> could be solved in a library on top of it or in findup-sync.

I'm open to any of those solutions, but my preference would be a separate library if not here - since this will need to be done in other libraries I maintain.

doowb commented:
> my preference would be a separate library if not here

I think that's the best approach (along with some of the other comments here).
I'll put together a repo that will mainly do what this PR is doing:
- if filepath exists return filepath
- if filepath does not exist
  - **default** return null
  - **not win32 platform | opt-in flag** compare against files in directory

So we can either always do the checking if the initial check returns `false` or base the check on the platform or an option. (I'd rather do an option and let a user decide).

We can also have an optional function for doing the filepath comparisons (this would basically default to what's in here).

phated commented:
👍  but it will be **_not_ win32 platform | opt-in flag**, i think

jonschlinkert commented:
> but it will be not win32 platform | opt-in flag, i think

agreed, that makes sense

doowb commented:
>  but it will be not win32 platform | opt-in flag, i think

Sorry, yeah. I edited the comment.

Also, I edited to reflect that the module will not be recursing through directories. I'll only check the directory the given filepath is in (by dirname on filepath like it's doing now).

The module will only detect that a given file or directory exists and return the filepath that is detected (so users can get the correctly cased filepath on case sensitive systems).

Given that, I think `detect-file` is a good name. I've already created the repo with code and tests. I'll get it published in a few minutes for everyone to take a look at.

jonschlinkert commented:
> I think detect-file is a good name.

makes sense to me, since it's not really finding/searching, just verifying that the file or directory exists. 

fwiw I'll be using that lib in [find-file-up](https://github.com/jonschlinkert/find-file-up) as well.

doowb commented:
I created [detect-file](https://github.com/doowb/detect-file) which handles detecting if a file or directory exists and returning the resolved filepath. Set the `nocase` option to `true` to force case-insensitive checking.

I updated `findup-sync` on [this branch on my fork](https://github.com/doowb/node-findup-sync/tree/fs-exists-sync) to use `detect-file` and set `nocase: process.platform !== 'win32'`. This will only trigger the directory reading when a file is not initial found and we're on a potentially case sensitive file system.

I updated some of the tests and [ran them on travis-ci](https://travis-ci.org/doowb/node-findup-sync) on linux and osx. I also logged into a ubuntu box and tested `gulp 3` and `gulp-cli` running `gulp 3` and `gulp 4` while replacing `findup-sync` with my fork in `liftoff` dependencies. I used `Gulpfile.js` to ensure the case difference was working.

If this looks good. I can do a PR to `findup-sync` with the squashed updates from my fork.

phated commented:
👍  I can't think of a way someone would be able to install Windows on a case-sensitive filesystem but Liftoff just uses `nocase: true` so that option can be passed directly through to `findup-sync` to make it support that option the way it used to and Liftoff shouldn't need any changes.

jonschlinkert commented:
> Liftoff just uses nocase: true so that option can be passed directly through to findup-sync

great, that's exactly what we were thinking. we tried to sync up options to make this smooth. fwiw sorry this took so long. the solutions we came up with didn't sit right with me, I had to think about it

jonschlinkert commented:
clicked the wrong button! pizza got here right when I was closing lol
